Libplanet/Net/Swarm.cs

```diff
@@ -752,84 +794,58 @@ public string TraceTable()
-            }
-            catch (TimeoutException e)
-            {
-                _logger.Error(e, "Timeout occurred during SendMessageAsync().");
-                throw;
-            }
-            catch (Exception e)
-            {
-                _logger.Error(
-                    e,
-                    $"An unexpected exception occurred during {nameof(SendMessageAsync)}(). {e}");
-                throw;
-            }
-            finally
-            {
-                if (_dealers.ContainsKey(peer))
-                {
-                    _dealers[peer] = (DateTimeOffset.UtcNow, _dealers[peer].Item2);
-                }
-                else
-                {
-                    throw new Exception("Dealer not exists.");
-                }
-            }
+            await SendMessageWithReplyAsync(peer, message, TimeSpan.FromSeconds(3), 0);
         }
 
         internal async Task<Message> SendMessageWithReplyAsync(
             BoundPeer peer,
             Message message,
             TimeSpan? timeout,
             CancellationToken cancellationToken)
+        {
+            IEnumerable<Message> replies =
+                await SendMessageWithReplyAsync(peer, message, timeout, 1, cancellationToken);
+            Message reply = replies.First();
+            ValidateSender(reply.Remote);
+
+            return reply;
+        }
+
+        internal async Task<IEnumerable<Message>> SendMessageWithReplyAsync(
+            BoundPeer peer,
+            Message message,
+            TimeSpan? timeout,
+            int expectedResponses,
+            CancellationToken cancellationToken = default(CancellationToken)
+        )
         {
             if (!(_turnClient is null))
             {
                 await CreatePermission(peer);
             }
 
+            Guid reqId = Guid.NewGuid();
             try
             {
-                using (var dealer = new DealerSocket(ToNetMQAddress(peer)))
-                {
-                    _logger.Debug($"Trying to send [{message}] to [{peer.Address.ToHex()}]...");
-
-                    await dealer.SendMultipartMessageAsync(
-                        message.ToNetMQMessage(_privateKey, AsPeer),
-                        timeout: timeout,
-                        cancellationToken: cancellationToken);
-
-                    _logger.Debug($"Message sent, waiting for reply...");
-
-                    NetMQMessage raw = await dealer.ReceiveMultipartMessageAsync(
-                        timeout: timeout,
-                        cancellationToken: cancellationToken);
+                _logger.Verbose("Adding request[{0}] to queue...", reqId);
+                var tcs = new TaskCompletionSource<IEnumerable<Message>>();
+                await _requests.AddAsync(
+                    new MessageRequest
+                    {
+                       Id = reqId,
+                       Message = message,
+                       Peer = peer,
+                       Timeout = timeout,
```

Comment:
Is it okay if this field is `null`?

Reply:
Yes, if `MessageRequest.Timeout` is null, it means infinite waiting. 

or, Should we prohibit infinite waiting?

Reply:
Seems okay as it is.